### PR TITLE
core: expose rectangle intersection to bindings

### PR DESCRIPTION
**Pull request by paroj**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/core/include/opencv2/core/types.hpp [hunk 1]
```diff
@@ -2017,6 +2017,15 @@ double jaccardDistance(const Rect_<_Tp>& a, const Rect_<_Tp>& b) {
...
 }
 
+/** @brief Finds out if there is any intersection between two rectangles
+ * 
+ * mainly useful for language bindings
+ * @param rect1 First rectangle
+ * @param rect2 Second rectangle
+ * @return the area of the intersection
+ */
+CV_EXPORTS_W inline double rectangleIntersection(const Rect2d& a, const Rect2d& b) { return (a & b).area(); }
```
Comment (mshabunin) [inline modules/core/include/opencv2/core/types.hpp:2027]: Shouldn't we also add this function for `cv::Rect`? Will there be a conflict if the have the same name?

Comment (alalek) [reply modules/core/include/opencv2/core/types.hpp:2027]: > rectangleIntersection

Name is not accurate. It returns "area".

---

There is no `cv::Rect` automatic wrapper in bindings.
We could add helper Python class for Rect support here: https://github.com/opencv/opencv/tree/4.x/modules/core/misc/python/package

Comment (mshabunin) [reply modules/core/include/opencv2/core/types.hpp:2027]: There is a wrapper for `cv::Rect` in common code: https://github.com/opencv/opencv/blob/0ac67ee8d92e3fef8bb78e774cc004b8cdba180e/modules/python/src2/cv2_convert.cpp#L674-L687

Comment (paroj) [reply modules/core/include/opencv2/core/types.hpp:2027]: > Shouldn't we also add this function for cv::Rect? Will there be a conflict if the have the same name?

currently int rects will be converted to double by parseSequence, so they do work. An int overload would be handled by the wrapper just fine, but I doubt there would be any performance benefits (using python)

> Name is not accurate. It returns "area".

tried to follow rotatedRectangleIntersection here, which also returns the intersection area. Then there is intersectConvexConvex. What name would you suggest? rectangleIntersectionArea sounds rather verbose..

**Review by mshabunin — APPROVED** [review]

File: modules/core/include/opencv2/core/types.hpp (hunk 1 above)
Comment (paroj) [reply modules/core/include/opencv2/core/types.hpp:2027]: @alalek the naming is the only outstanding issue here, right? Any suggestions?

**Comment by fengyuentau** [comment]
Could we expose the rectangle union as well? I could see there is a great demand when it comes to calculatoing IoU (intersection over union) in object detection. And these two APIs could help simplify the calculation.

**Comment by asmorkalov** [comment]
@paroj Please fix warnings on CI and I'll merge the patch:
```
modules/core/include/opencv2/core/types.hpp:2021: trailing whitespace.
+ * 
```

**Pull request merged**